### PR TITLE
Replace Environment.Exit with ProcessExitSource and cancelation token

### LINKED ISSUE #5298: Occasional problems with CI
The failing build run link https://github.com/NethermindEth/nethermind/actions/runs/4818823246/jobs/8581269720

image

```
Passed!  - Failed:     0, Passed:    29, Skipped:     0, Total:    29, Duration: 26 s - Nethermind.Runner.Test.dll (net7.0)
Test Run Aborted with error System.Exception: One or more errors occurred.
 ---> System.Exception: Unable to read beyond the end of the stream.
   at System.IO.BinaryReader.Read7BitEncodedInt()
   at System.IO.BinaryReader.ReadString()
   at Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.LengthPrefixCommunicationChannel.NotifyDataAvailable()
   at Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.TcpClientExtensions.MessageLoopAsync(TcpClient client, ICommunicationChannel channel, Action`1 errorHandler, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---.
Error: Process completed with exit code 1.
```

### PR DESCRIPTION
**Pull request by LukaszRozmej**
Potentially Fixes #5298

## Changes

- Removes all Environment.Exit and replaces them by using wrapped cancellation token source

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Runner/Program.cs [hunk 1]
```diff
@@ -210,23 +210,15 @@ private static void Run(string[] args)
...
-                    if (e is IExceptionWithExitCode withExit)
-                    {
-                        exitCode = withExit.ExitCode;
-                    }
-                    else
-                    {
-                        exitCode = ExitCodes.GeneralError;
-                    }
-                    _processCloseCancellationSource.Cancel();
+                    _processExitSource.ExitCode = e is IExceptionWithExitCode withExit ? withExit.ExitCode : ExitCodes.GeneralError;
```
Comment (Scooletz) [inline src/Nethermind/Nethermind.Runner/Program.cs:213]: Now setting the `ExitCode` won't cances the source. Previously, there was `Cancel` call. Is this a problem? If no, can you elaborate why?

File: src/Nethermind/Nethermind.Core/Extensions/CancellationTokenExtensions.cs [hunk 2]
```diff
@@ -38,8 +38,11 @@ public static void CancelDisposeAndClear(ref CancellationTokenSource? cancellati
                 {
                     source.Cancel();
                     source.Dispose();
+                    return true;
```
Comment (Scooletz) [inline src/Nethermind/Nethermind.Core/Extensions/CancellationTokenExtensions.cs:41]: A nitpick: the `CompareExchange` above can be replaced with `Exchange` as you always get the previous value.

**Review by Scooletz — COMMENTED** [review]
One nitpick and one more important comment.

File: src/Nethermind/Nethermind.Runner/Program.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Runner/Program.cs]: Right got confused.

File: src/Nethermind/Nethermind.Core/Extensions/CancellationTokenExtensions.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Core/Extensions/CancellationTokenExtensions.cs:41]: Well depends, if someone would have swapped it for something else by then, then we could break that logic.

**Comment by LukaszRozmej** [comment]
Unfortunately didn't fix:

Test Run Aborted with error System.Exception: One or more errors occurred.
 ---> System.Exception: Unable to read beyond the end of the stream.
   at System.IO.BinaryReader.Read7BitEncodedInt()
   at System.IO.BinaryReader.ReadString()
   at Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.LengthPrefixCommunicationChannel.NotifyDataAvailable()
   at Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.TcpClientExtensions.MessageLoopAsync(TcpClient client, ICommunicationChannel channel, Action`1 errorHandler, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---.
Error: Process completed with exit code 1.

**Pull request merged**